relay/chunk_manager.go

```diff
@@ -0,0 +1,165 @@
+	ctx    context.Context
+	logger logging.Logger
+
+	// metadataCache is an LRU cache of blob metadata. Blobs that do not belong to one of the relay shards
+	// assigned to this server will not be in the cache.
+	frameCache cache.CachedAccessor[blobKeyWithMetadata, []*encoding.Frame]
+
+	// chunkReader is used to read chunks from the chunk store.
+	chunkReader chunkstore.ChunkReader
+
+	// pool is a work pool for managing concurrent requests. Used to limit the number of concurrent
+	// requests.
+	pool *errgroup.Group
+}
+
+// blobKeyWithMetadata attaches some additional metadata to a blobKey.
+type blobKeyWithMetadata struct {
+	blobKey  v2.BlobKey
+	metadata blobMetadata
+}
+
+func (m *blobKeyWithMetadata) Compare(other *blobKeyWithMetadata) int {
+	return bytes.Compare(m.blobKey[:], other.blobKey[:])
+}
+
+// newChunkManager creates a new chunkManager.
+func newChunkManager(
+	ctx context.Context,
+	logger logging.Logger,
+	chunkReader chunkstore.ChunkReader,
+	cacheSize int,
+	workPoolSize int) (*chunkManager, error) {
+
+	pool := &errgroup.Group{}
+	pool.SetLimit(workPoolSize)
+
+	server := &chunkManager{
+		ctx:         ctx,
+		logger:      logger,
+		chunkReader: chunkReader,
+		pool:        pool,
+	}
+
+	c, err := cache.NewCachedAccessor[blobKeyWithMetadata, []*encoding.Frame](cacheSize, server.fetchFrames)
+	if err != nil {
+		return nil, err
+	}
+	server.frameCache = c
+
+	return server, nil
+}
+
+// frameMap is a map of blob keys to frames.
+type frameMap map[v2.BlobKey][]*encoding.Frame
+
+// GetFrames retrieves the frames for a blob.
+func (s *chunkManager) GetFrames(ctx context.Context, mMap *metadataMap) (*frameMap, error) {
+
+	keys := make([]*blobKeyWithMetadata, 0, len(*mMap))
+	for k, v := range *mMap {
+		keys = append(keys, &blobKeyWithMetadata{blobKey: k, metadata: *v})
+	}
+
+	fMap := make(frameMap, len(keys))
+	hadError := atomic.Bool{}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(keys))
+
+	for _, key := range keys {
+		boundKey := key
+		s.pool.Go(func() error {
+
+			defer wg.Done()
+
+			frames, err := s.frameCache.Get(*boundKey)
+			if err != nil {
+				s.logger.Error("Failed to get frames for blob %v: %v", boundKey.blobKey, err)
+				hadError.Store(true)
+			} else {
+				fMap[boundKey.blobKey] = *frames
+			}
+
+			return nil
+		})
+	}
+
+	wg.Wait()
+
+	return &fMap, nil
+}
+
+// fetchFrames retrieves the frames for a single blob.
+func (s *chunkManager) fetchFrames(key blobKeyWithMetadata) (*[]*encoding.Frame, error) {
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	var proofs []*encoding.Proof
+	var proofsErr error
+
+	s.pool.Go(
+		func() error {
+			defer wg.Done()
+			proofs, proofsErr = s.chunkReader.GetChunkProofs(s.ctx, key.blobKey)
+			return nil
+		})
+
+	fragmentInfo := &encoding.FragmentInfo{
+		TotalChunkSizeBytes: key.metadata.totalChunkSizeBytes,
+		FragmentSizeBytes:   key.metadata.fragmentSizeBytes,
+	}
+
+	coefficients, err := s.chunkReader.GetChunkCoefficients(s.ctx, key.blobKey, fragmentInfo)
```

Comment:
should this request also execute from the pool?

Reply:
`GetChunkCoefficients()` fans out into the pool of workers used by the S3 client. It does this since we break the files into tiny fragments for parallel read/write. So even if we have a very large number of parallel requests, the total number of active S3/dynamoDB connections has a firm upper limit controlled by us.

Reply:
Ah makes sense!